### PR TITLE
(CPR-515) Allow separate tasks for install vs upgrade

### PR DESCRIPTION
**Pull request by shrug**
Currently, the postinstall tasks from the packaged project
are executed unconditionally, both on install as well as
upgrades. This can cause issues, as in the case with puppetserver
potentially overwriting config values if the user has changed them
from the package defaults.

This adds an addtional set of tasks that can be executed in the
package post-install script, that are only executed on initial
package install. This leaves the current behavior untouched, so
packages depending on the current behavior can be unmodified.

### CODE REVIEW

**Comment by shrug** [comment]
We could add another section for upgrades, add them to other scriptlets, etc. This is the minimum needed to satisfy the current requirement.

**Comment by shrug** [comment]
Another way forward would be to wrap each task in, for example, https://github.com/shrug/puppet-server/blob/a523c1ca4ab9924fc8f886ece30cf193e4be6d7c/resources/ext/ezbake.conf#L24-L34 in conditionals depending on the requirements. That seems less maintainable, but we already do that in a limited fashion in PuppetDB, for example.

File: resources/puppetlabs/lein-ezbake/template/global/install.sh.erb [hunk 1]
```diff
@@ -327,22 +327,38 @@ function task_preinst_deb {
...
 <% EZBake::Config[:debian][:additional_postinst].each do |cmd| -%>
     <%= cmd %>
 <% end -%>
 }
 
-# RPM based post-installation tasks.
+# Debian based install post-installation tasks.
+function task_postinst_deb_install {
+    task postinst_permissions
+<% EZBake::Config[:debian][:additional_postinst_install].each do |cmd| -%>
```
Comment (mwaggett) [inline resources/puppetlabs/lein-ezbake/template/global/install.sh.erb]: Do you need a null command for this one too?

File: resources/puppetlabs/lein-ezbake/template/global/ext/debian/postinst.erb [hunk 2]
```diff
@@ -2,6 +2,11 @@
 
 /opt/puppetlabs/server/apps/<%= EZBake::Config[:real_name] -%>/scripts/install.sh postinst_deb
 
+# Set some configuration from the project on install
+if [ -z "$2" ]; then
```
Comment (mwaggett) [inline resources/puppetlabs/lein-ezbake/template/global/ext/debian/postinst.erb]: Where's the value of `$2` coming from?
also, is it meant to be in quotes?

File: resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb [hunk 3]
```diff
@@ -223,7 +223,13 @@ fi
 <% end -%>
 
 %post
+# Unconditional post-install tasks
 %{_app_prefix}/scripts/install.sh postinst_redhat
+
+# Only run these on initial install, not ugrades
+if [ "$1" = "1" ]; then
```
Comment (mwaggett) [inline resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb]: same questions about values and quotes

File: resources/puppetlabs/lein-ezbake/template/global/ext/debian/postinst.erb (hunk 2 above)
Comment (underscorgan) [reply resources/puppetlabs/lein-ezbake/template/global/ext/debian/postinst.erb]: `$2` is a variable available in debian packaging, set to the previous version if this is an upgrade, not set if it's a fresh install. Yes, it needs to be in quotes, we're testing to see whether or not this is an empty string

File: resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb (hunk 3 above)
Comment (underscorgan) [reply resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb]: redhat does this a little differently, but still a variable that lets you know whether this is an install/upgrade/etc. See the chart at https://fedoraproject.org/wiki/Packaging:Scriptlets. You need to quote it for the bash comparison

File: resources/puppetlabs/lein-ezbake/template/global/install.sh.erb (hunk 1 above)
Comment (underscorgan) [reply resources/puppetlabs/lein-ezbake/template/global/install.sh.erb]: Nope, since there's the task postinst_permissions in this function already. That's only needed if you're going to end up with an empty block

**Comment by underscorgan** [comment]
Assuming this works as expected on test, 👍

**Comment by shrug** [comment]
Test build: http://builds.delivery.puppetlabs.net/puppetserver/5.1.6.SNAPSHOT.2018.02.07T1415/

File: resources/puppetlabs/lein-ezbake/template/global/install.sh.erb [hunk 4]
```diff
@@ -327,22 +327,38 @@ function task_preinst_deb {
...
     task postinst_permissions
 <% EZBake::Config[:debian][:additional_postinst].each do |cmd| -%>
     <%= cmd %>
 <% end -%>
 }
 
-# RPM based post-installation tasks.
+# Debian based install post-installation tasks.
+function task_postinst_deb_install {
+    task postinst_permissions
```
Comment (haus) [inline resources/puppetlabs/lein-ezbake/template/global/install.sh.erb]: Will calling postinst_permissions twice cause problems?

Comment (shrug) [reply resources/puppetlabs/lein-ezbake/template/global/install.sh.erb]: oh yeah we should drop that. copy-pasta

File: resources/puppetlabs/lein-ezbake/template/global/install.sh.erb [hunk 5]
```diff
@@ -327,22 +327,37 @@ function task_preinst_deb {
...
     <%= cmd %>
 <% end -%>
 }
 
-# RPM based post-installation tasks.
+# Debian based install post-installation tasks.
+function task_postinst_deb_install {
+<% EZBake::Config[:debian][:additional_postinst_install].each do |cmd| -%>
+    <%= cmd %>
+<% end -%>
```
Comment (underscorgan) [inline resources/puppetlabs/lein-ezbake/template/global/install.sh.erb]: You need a null command in here now that you've removed the postinst_permissions

File: resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb [hunk 6]
```diff
@@ -223,7 +223,13 @@ fi
...
+  %{_app_prefix}/scripts/install.sh postinst_redhat_install
+fi
```
Comment (underscorgan) [inline resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb]: With the FPM stuff you actually want to be updating this file, https://github.com/puppetlabs/ezbake/blob/master/resources/puppetlabs/lein-ezbake/template/global/ext/redhat/postinst.erb

The spec file is only used for legacy builds. I want to burn that down for exactly this reason Soon, but it's not there yet

sad

Comment (shrug) [reply resources/puppetlabs/lein-ezbake/template/global/ext/redhat/ezbake.spec.erb]: oh cool, ok I was wondering how the standalone files were being used.

**Comment by shrug** [comment]
Updated to address comments

**Comment by shrug** [comment]
This now works as expected for RPM - installs set the config variables, upgrades do not overwrite them.

**Comment by shrug** [comment]
Also works on debian. 

Tested with http://builds.delivery.puppetlabs.net/puppetserver/5.1.6.SNAPSHOT.2018.02.07T1610/
